### DISCUSSION TITLE
IKEv2: macOS & strongSwan in 2021

### DISCUSSION THREAD

**Discussion by ewann** [strongswan/strongswan#777] [Show and tell]
macOS version: `12.0.1` (perhaps earlier versions, not extensively tested)
strongSwan version: `5.7.2` (perhaps earlier and / or later versions, not extensively tested)
IKEv2 certificate based connections

tldr:

- Re-keying _does_ work out of the box and interoperate sufficiently well
- macOS still doesn't react to rfc4478 `AUTH_LIFETIME` messages in an interoperable way - strongSwan `ikelifetime` (if using `ipsec.conf`) is significant to the upper time limit of a connection, and defaults to ~3 hours.
- Split DNS (for split tunneling) will get configured on macOS side if `attr.25` is configured in `charon/attr.conf`
- Connection configuration on macOS side can be done using the UI - no settings necessary that require Apple Configurator 2. Equally, there is nothing that seems to prevent using AC2, but consider if / why you want to constrain Security Association parameters here. If you don't know why, don't. - https://wiki.strongswan.org/projects/strongswan/wiki/SecurityRecommendations
- `charondebug="mgr 2, ike 2, knl 1, cgf 2"` (if using `ipsec.conf`) could be a good logging level to start with if needing to debug the `AUTH_LIFETIME` issue 

I recently needed to investigate & resolve these issues. While there is various documentation at the strongSwan site, there is also 3rd party opinion elsewhere on the internet about how to resolve some of these issues. Over time some of this material has become conflicting, possibly invalid. Hopefully by commenting this information here, it may help somebody else accelerate achievement of a satisfactory (to them) solution.

https://wiki.strongswan.org/projects/strongswan/wiki/AppleClients 
- it no longer appears necessary to make the mentioned ike/esp changes, at least for macOS `12.0.1`
- you may have security requirements that prevent you from completely disabling `reauth`. As mentioned see `ikelifetime` in https://wiki.strongswan.org/projects/strongswan/wiki/ExpiryRekey for `ipsec.conf` (or `reauth_time` if using swanctl.conf) if you need to adjust the time period between `reauth` attempts. The effect on macOS is that it will continue to be disconnected after this time period, but you can choose your own compromise between user experience and security.

https://wiki.strongswan.org/projects/strongswan/wiki/AttrPlugin
- attribute 25 is not defined, however it can be specified by number (https://datatracker.ietf.org/doc/html/rfc8598)
- split tunnel, dns domains example: `/etc/strongswan/strongswan.d/charon/attr.conf`
```
attr {
    dns = <some.dns.server> # eg: 192.168.0.3
    25 = <some.comma.seperated>, <values.tld> #eg: internal.example.com, amazonaws.com
    load = yes
}
```
- based on the above, after connecting on macOS `scutil --dns` should show similar to:
```
DNS configuration

resolver #1   
  search domaion[0] : internal.example.com
  search domaion[1] : amazonaws.com
  nameserver[0] : <your local nameserver IP outside the vpn>
  if_index...

resolver #2 
  domain...
```

**Comment by kwohlfahrt** [comment]
A small gotcha `nslookup` does not use the same DNS stack as many other MacOS utilities. I was trying to debug why my internal domain still couldn't be resolved, but it turns out that while `nslookup` failed:

```
[I] kai@Kais-MBP ~/D/c/c/i/vpn (kwohlfahrt/private-dns) [68]> nslookup test.kai.i.charmtx.com
Server:		192.168.178.1
Address:	192.168.178.1#53

** server can't find test.kai.i.charmtx.com: NXDOMAIN
```

All other commands succeeded:
```
[I] kai@Kais-MBP ~/D/c/c/i/vpn (kwohlfahrt/private-dns) [1]> ping -c3 test.kai.i.charmtx.com
PING test.kai.i.charmtx.com (172.25.129.11): 56 data bytes
64 bytes from 172.25.129.11: icmp_seq=0 ttl=59 time=87.061 ms
64 bytes from 172.25.129.11: icmp_seq=1 ttl=59 time=86.222 ms
64 bytes from 172.25.129.11: icmp_seq=2 ttl=59 time=87.204 ms

--- test.kai.i.charmtx.com ping statistics ---
3 packets transmitted, 3 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 86.222/86.829/87.204/0.433 ms
```
